### DISCUSSION TITLE
Limit the number of messages being sent to channel

### DISCUSSION THREAD

**Discussion by ahmednasir91** [celery/kombu#2159] [Q&A]
So we are facing a situation where RabbitMQ is closing the connect of celery queue on write time and basically it comes with `PRECONDITION_FAILED message size is greater than 128MB` (the default in RMQ).

We tried to override the `kombu.Producer.publish` to catch which message is causing the issue, looks like the tasks we have payload size each of 22MB but when they start executing as chunks in a chord there is a message sent which contains multiple tasks and hence the timeout.

How can i limit batching of these messages in kombu or is this happening in celery itself?

**Comment by TayfurYldz** [comment]
The 128 MB frame is not Kombu batching several independent Producer.publish calls together. Kombu normally publishes each Celery message separately; its transport-level batch API is opt-in and is not what Celery chords use here.

For a chord/group, Celery serializes orchestration metadata and, depending on backend/primitive, may create messages containing signatures/results for multiple tasks. So if your individual task payload is already ~22 MB, a group/chord message can cross RabbitMQ's max_message_size even though no individual business payload does.

The first diagnostic should therefore be at the Celery canvas layer, not by overriding Kombu batching. Log the routing key/task name plus the serialized body size immediately before publish and identify which Celery task/message type is ~128 MB. If it is the chord/group orchestration message, there is no Kombu "max tasks per AMQP message" knob that can safely split an already serialized Celery protocol message after the fact.

The robust fix is to stop putting large data in task signatures/results. Store the large payload externally (object store/database/etc.) and pass compact identifiers between tasks. Also inspect the result backend/chord path, because chord aggregation can make large results compound just like large arguments.

Raising RabbitMQ's max_message_size may postpone the failure, but six 22 MB payloads already put you in a range where retries, replication, memory pressure and serialization become expensive. I would treat the broker exception as exposing an oversized Celery message and reduce the protocol payload rather than trying to batch-limit it in Kombu.
